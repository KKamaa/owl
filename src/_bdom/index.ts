import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.ops.mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  block1.ops.patch(block1, block2);
}

export function remove(block: Block) {
  // beforeRemove(block);
  block.ops.remove(block);
}
