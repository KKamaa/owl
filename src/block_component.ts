import { Block } from "./bdom";
import { makeChildFiber, __internal__destroyed } from "./fibers";
import { OwlNode } from "./owl_node";
import { STATUS } from "./status";

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

export class BComponent extends Block {
  node: OwlNode;
  parentClass?: any;
  classTarget?: HTMLElement;

  /**
   *
   * @param name
   * @param props
   * @param key
   * @param owner the component in which the component was defined
   * @param parent the actual parent (may be different in case of slots)
   */
  constructor(name: string, props: any, key: string, owner: any, parent: any) {
    super();
    const parentNode: OwlNode = parent.__owl__;
    let node: OwlNode | undefined = parentNode.children[key];
    if (node && node.status < STATUS.MOUNTED) {
      node.destroy();
      // delete parentNode.children[key];
      node = undefined;
    }
    const parentFiber = parentNode.fiber!;
    if (node) {
      // update
      const fiber = makeChildFiber(node, parentFiber);
      if (node.willPatch.length) {
        parentFiber.root.willPatch.push(fiber);
      }
      if (node.patched.length) {
        parentFiber.root.patched.push(fiber);
      }
      node.updateAndRender(props, fiber);
    } else {
      // new component
      const components = owner.constructor.components;
      const C = components[name];
      node = new OwlNode(parentNode.app, C, props);
      parentNode.children[key] = node;
      const fiber = makeChildFiber(node, parentNode.fiber!);
      if (node.mounted.length) {
        parentFiber.root.mounted.push(fiber);
      }
      node.initiateRender(fiber);
    }
    this.node = node;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.node.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    const node = this.node;
    const bdom = node.fiber!.bdom!;
    node.bdom = bdom;
    bdom.mountBefore(anchor);
    if (this.parentClass) {
      this.parentClass = this.parentClass.trim().split(/\s+/);
      const el = this.firstChildNode();
      if (el instanceof HTMLElement) {
        this.addClass(el);
      }
    }
    node.status = STATUS.MOUNTED;
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  moveBefore(anchor: ChildNode) {
    this.node.bdom!.moveBefore(anchor);
  }

  addClass(el: HTMLElement) {
    this.classTarget = el;
    for (let cl of this.parentClass) {
      el.classList.add(cl);
    }
  }

  removeClass(el: HTMLElement) {
    for (let cl of this.parentClass) {
      el.classList.remove(cl);
    }
  }

  patch() {
    const node = this.node;

    node.bdom!.patch(node!.fiber!.bdom!);
    if (this.parentClass) {
      const el = this.firstChildNode();
      if (el !== this.classTarget) {
        if (el && this.classTarget) {
          this.removeClass(this.classTarget);
          this.addClass(el as any);
        } else if (el) {
          this.addClass(el as any);
        } else {
          this.removeClass(this.classTarget!);
          this.classTarget = undefined;
        }
      }
    }
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  beforeRemove() {
    visitRemovedNodes(this.node);
  }

  remove() {
    const bdom = this.node.bdom!;
    bdom.remove();
  }
}

function visitRemovedNodes(node: OwlNode) {
  if (node.status === STATUS.MOUNTED) {
    const component = node.component;
    for (let cb of node.willUnmount) {
      cb.call(component);
    }
  }
  for (let child of Object.values(node.children)) {
    visitRemovedNodes(child);
  }
  node.status = STATUS.DESTROYED;
  if (node.destroyed.length) {
    __internal__destroyed.push(node);
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, owner: any, parent: any) {
    super(name, props, key, owner, parent);
    this.handlers = new Array(handlers);
  }
  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }
  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.node.component.el!;
      el.addEventListener(eventType, (ev: Event) => {
        const info = this.handlers![i];
        const [, callback] = info;
        callback(ev);
      });
    }
  }
}
