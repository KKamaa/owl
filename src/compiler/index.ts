import { setupBlockDom } from "./helpers";
import { QWebCompiler, TemplateFunction } from "./qweb_compiler";
export { Template, TemplateFunction } from "./qweb_compiler";

setupBlockDom();

export function compileTemplate(template: string, name?: string): TemplateFunction {
  const compiler = new QWebCompiler(template, name);
  return compiler.compile();
}
