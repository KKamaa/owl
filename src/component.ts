import type { BNode } from "./b_node";

export class Component {
  static template: string = "";

  props: any;
  env: any;
  __owl__: BNode;

  constructor(props: any, env: any, node: BNode) {
    this.props = props;
    this.env = env;
    this.__owl__ = node;
  }
  get el(): HTMLElement | null {
    const node = this.__owl__;
    return node.bdom ? (node.bdom.el as any) : null;
  }

  setup() {}

  render(): Promise<void> {
    return this.__owl__.render();
  }
}
