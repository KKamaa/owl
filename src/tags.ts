// -----------------------------------------------------------------------------
//  Global templates
// -----------------------------------------------------------------------------

export const globalTemplates: { [key: string]: string } = {};

export function xml(strings: TemplateStringsArray, ...args: any[]) {
  const name = `__template__${xml.nextId++}`;
  const value = String.raw(strings, ...args);
  globalTemplates[name] = value;
  return name;
}

xml.nextId = 1;
