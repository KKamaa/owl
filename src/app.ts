import { Blocks as BaseBlocks } from "./bdom";
import { compileTemplate, Template } from "./compiler/index";
import { globalTemplates } from "./tags";
import { BComponent, BComponentH } from "./block_component";
import type { Component } from "./component";
import { OwlNode } from "./owl_node";
import { Scheduler } from "./scheduler";
import { UTILS } from "./template_utils";

const Blocks = {
  ...BaseBlocks,
  BComponent,
  BComponentH,
};

// -----------------------------------------------------------------------------
//  TemplateSet
// -----------------------------------------------------------------------------

export class TemplateSet {
  rawTemplates: { [name: string]: string } = Object.create(globalTemplates);
  templates: { [name: string]: Template } = {};
  utils: typeof UTILS;

  constructor() {
    const call = (subTemplate: string, ctx: any, parent: any) => {
      const template = this.getTemplate(subTemplate);
      return template(ctx, parent);
    };

    this.utils = Object.assign({}, UTILS, { call });
  }

  addTemplate(name: string, template: string, options: { allowDuplicate?: boolean } = {}) {
    if (name in this.rawTemplates && !options.allowDuplicate) {
      throw new Error(`Template ${name} already defined`);
    }
    this.rawTemplates[name] = template;
  }

  getTemplate(name: string): Template {
    if (!(name in this.templates)) {
      const rawTemplate = this.rawTemplates[name];
      if (rawTemplate === undefined) {
        throw new Error(`Missing template: "${name}"`);
      }
      const templateFn = compileTemplate(rawTemplate, name);
      const template = templateFn(Blocks, this.utils);
      this.templates[name] = template;
    }
    return this.templates[name];
  }
}

export class App<T extends typeof Component = any> extends TemplateSet {
  Root: T;
  props: any;
  env: any = {};
  scheduler = new Scheduler(window.requestAnimationFrame.bind(window));
  root: OwlNode | null = null;
  dev: boolean = true;
  
  constructor(Root: T, props?: any) {
    super();
    this.Root = Root;
    this.props = props;
  }

  configure(params: any) {
    if (params.env) {
      this.env = params.env;
    }
  }

  mount(target: HTMLElement): Promise<InstanceType<T>> {
    if (!(target instanceof HTMLElement)) {
      throw new Error("Cannot mount component: the target is not a valid DOM element");
    }
    if (!document.body.contains(target)) {
      throw new Error("Cannot mount a component on a detached dom node");
    }
    const node = new OwlNode(this, this.Root, this.props);
    this.root = node;
    return node.mount(target);
  }

  destroy() {
    if (this.root) {
      this.root.destroy();
    }
  }
}
