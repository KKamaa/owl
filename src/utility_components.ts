import { Component } from "./component";
import type { OwlNode } from "./owl_node";
import { xml } from "./tags";

export class NoUpdate extends Component {
  static template = xml`<t t-slot="default"/>`;

  constructor(props: any, env: any, node: OwlNode) {
    super(props, env, node);
    node._render = function (fiber: any) {
      if (!this.bdom) {
        this.bdom = this.renderFn();
        this.bdom!.patch = () => {};
      }
      fiber.bdom = this.bdom;
      fiber.root.counter--;
    };
  }
}
