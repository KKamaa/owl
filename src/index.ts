import { App } from "./app";
import { Component } from "./component";

export { App, Component };

export async function mount<T extends typeof Component>(
  C: T,
  target: HTMLElement
): Promise<InstanceType<T>> {
  const app = new App(C);
  return app.mount(target);
}

export { xml } from "./tags";
export { useState } from "./reactivity";
export { useRef } from "./refs";
export { EventBus } from "./event_bus";

export {
  onWillStart,
  onMounted,
  onBeforeUnmount,
  onWillUpdateProps,
  onBeforePatch,
  onPatched,
} from "./lifecycle_hooks";

export const __info__ = {};
