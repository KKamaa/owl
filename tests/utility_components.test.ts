import { Component, mount, useState } from "../src";
import { xml } from "../src/tags";
import { NoUpdate } from "../src/utility_components";
import { makeTestFixture, nextTick, snapshotEverything } from "./helpers";

let fixture: HTMLElement;

snapshotEverything();

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("NoUpdate", () => {
  test("prevent renderings from above ", async () => {
    class Child extends Component {
      static template = xml`<t t-esc="props.value"/>`;
    }
    class Test extends Component {
      static template = xml`
            <Child value="state.value"/>
            <NoUpdate> 
                <Child value="state.value"/>
            </NoUpdate>`;

      static components = { NoUpdate, Child };

      state = useState({ value: 1 });
    }

    const component = await mount(Test, fixture);

    expect(fixture.innerHTML).toBe("11");
    component.state.value = 2;
    await nextTick();
    expect(fixture.innerHTML).toBe("21");
  });
});
