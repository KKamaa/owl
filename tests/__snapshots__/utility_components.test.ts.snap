// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoUpdate prevent renderings from above  1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  return function template(ctx, parent = ctx) {
    const b1 = new BText(ctx['props'].value);
    return b1;
  }
}"
`;

exports[`NoUpdate prevent renderings from above  2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  const slot3 = ctx => parent => {
    const b2 = new BComponent(\`Child\`, {value: ctx['state'].value}, \`__4\`, ctx, parent);
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new BMulti(2);
    b1.children[0] = new BComponent(\`Child\`, {value: ctx['state'].value}, \`__1\`, ctx, parent);
    const b3 = b1.children[1] = new BComponent(\`NoUpdate\`, {}, \`__2\`, ctx, parent);
    b3.node.slots = {'default': slot3(ctx)};
    return b1;
  }
}"
`;

exports[`Portal Portal composed with t-slot 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BStatic {
    static el = elem(\`<div>child2</div>\`);
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`Portal Portal composed with t-slot 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  const slot2 = ctx => parent => {
    const b1 = callSlot(ctx, parent, 'default');
    return b1;
  }
  
  return function template(ctx, parent = ctx) {
    const b2 = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b2.node.slots = {'default': slot2(ctx)};
    return b2;
  }
}"
`;

exports[`Portal Portal composed with t-slot 3`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BComponentH(1, \`Child2\`, {}, \`__3\`, ctx, parent);
    b2.handlers[0] = ['custom', (e) => ctx.__owl__.component['_handled'](e)];
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    ctx = Object.create(ctx);
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Child\`, {}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal basic use of portal 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><span>1</span><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild.nextSibling;
    }
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<p>2</p>\`);
  }
  
  const slot2 = ctx => parent => {
    const b2 = new Block2();
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal conditional use of Portal (with sub Component) 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<p><owl-text/></p>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].val;
    return b1;
  }
}"
`;

exports[`Portal conditional use of Portal (with sub Component) 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BStatic {
    static el = elem(\`<span>1</span>\`);
  }
  
  const slot2 = ctx => parent => {
    const b3 = new BComponent(\`Child\`, {val: ctx['state'].val}, \`__3\`, ctx, parent);
    return b3;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new BMulti(2);
    const b2 = b1.children[0] = new Block1();
    if (ctx['state'].hasPortal) {
      const b4 = b1.children[1] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
      b4.node.slots = {'default': slot2(ctx)};
    }
    return b1;
  }
}"
`;

exports[`Portal conditional use of Portal 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BStatic {
    static el = elem(\`<span>1</span>\`);
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<p>2</p>\`);
  }
  
  const slot2 = ctx => parent => {
    const b3 = new Block2();
    return b3;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new BMulti(2);
    const b2 = b1.children[0] = new Block1();
    if (ctx['state'].hasPortal) {
      const b4 = b1.children[1] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
      b4.node.slots = {'default': slot2(ctx)};
    }
    return b1;
  }
}"
`;

exports[`Portal lifecycle hooks of portal sub component are properly called 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].val;
    return b1;
  }
}"
`;

exports[`Portal lifecycle hooks of portal sub component are properly called 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BComponent(\`Child\`, {val: ctx['state'].val}, \`__3\`, ctx, parent);
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    if (ctx['state'].hasChild) {
      const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
      b3.node.slots = {'default': slot2(ctx)};
    }
    return b1;
  }
}"
`;

exports[`Portal portal could have dynamically no content 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BMulti(1);
    if (ctx['state'].val) {
      const b3 = b2.children[0] = new Block2();
      b3.data[0] = ctx['state'].val;
    }
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b4 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b4.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal with child and props 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].val;
    return b1;
  }
}"
`;

exports[`Portal portal with child and props 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BComponent(\`Child\`, {val: ctx['state'].val}, \`__3\`, ctx, parent);
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal with dynamic body 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  class Block3 extends BStatic {
    static el = elem(\`<div/>\`);
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BMulti(2);
    if (ctx['state'].val) {
      const b3 = b2.children[0] = new Block2();
      b3.data[0] = ctx['state'].val;
    } else {
      const b4 = b2.children[1] = new Block3();
    }
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b5 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b5.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal with many children 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<div>1</div>\`);
  }
  
  class Block3 extends BStatic {
    static el = elem(\`<p>2</p>\`);
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BMulti(2);
    const b3 = b2.children[0] = new Block2();
    const b4 = b2.children[1] = new Block3();
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b5 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b5.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal with no content 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BMulti(1);
    if (false) {
      b2.children[0] = new BText('ABC');
    }
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal with only text as content 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BText('only text');
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal with target not in dom 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<div>2</div>\`);
  }
  
  const slot2 = ctx => parent => {
    const b2 = new Block2();
    return b2;
  }
  
  return (ctx, parent = ctx) => {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#does-not-exist'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal portal's parent's env is not polluted 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BStatic {
    static el = elem(\`<button>child</button>\`);
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`Portal portal's parent's env is not polluted 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BComponent(\`Child\`, {}, \`__3\`, ctx, parent);
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal with target in template (after portal) 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><span>1</span><owl-anchor/><div id=\\"local-target\\"/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild.nextSibling;
    }
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<p>2</p>\`);
  }
  
  const slot2 = ctx => parent => {
    const b2 = new Block2();
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#local-target'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal with target in template (before portal) 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><div id=\\"local-target\\"/><span>1</span><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild.nextSibling.nextSibling;
    }
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<p>2</p>\`);
  }
  
  const slot2 = ctx => parent => {
    const b2 = new Block2();
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#local-target'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;

exports[`Portal: UI/UX focus is kept across re-renders 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<input id=\\"target-me\\"/>\`);
    data = new Array(1);
    update() {
      this.updateAttr(this.el, \`placeholder\`, this.data[0]);
    }
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].val;
    return b1;
  }
}"
`;

exports[`Portal: UI/UX focus is kept across re-renders 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  const slot2 = ctx => parent => {
    const b2 = new BComponent(\`Child\`, {val: ctx['state'].val}, \`__3\`, ctx, parent);
    return b2;
  }
  
  return function template(ctx, parent = ctx) {
    const b1 = new Block1();
    const b3 = b1.children[0] = new BComponent(\`Portal\`, {target: '#outside'}, \`__1\`, ctx, parent);
    b3.node.slots = {'default': slot2(ctx)};
    return b1;
  }
}"
`;
